torch/distributed/algorithms/model_averaging/utils.py

```diff
@@ -32,5 +32,137 @@ def average_parameters(
 
     offset = 0
     for p in params_it2:
-        p.data = flat_params[offset : offset + p.numel()].view_as(p).type_as(p)
+        p.data = flat_params[offset: offset + p.numel()].view_as(p).type_as(p)
         offset += p.numel()
+
+
+class TensorBuffer:
+    """
+    Packs multiple tensors into one flat buffer for efficient
+    intra-worker communication.
+    refer to https://github.com/epfml/LocalSGD-Code/blob/3d4811d01673af205a00176f5389ed008a1ddb37/distributed_code/pcode/utils/tensor_buffer.py#L5
+    """
+
+    def __init__(self, tensors: List[torch.Tensor], use_cuda=True):
+        indices = [0]
+        for tensor in tensors:
+            new_end = indices[-1] + tensor.nelement()
+            indices.append(new_end)
+
+        self._start_idx = indices[:-1]
+        self._end_idx = indices[1:]
+        self._tensors_len = len(tensors)
+        self._tensors_sizes = [x.size() for x in tensors]
+
+        self.buffer = flatten(tensors, use_cuda=use_cuda)  # copies
+
+    def __getitem__(self, index):
+        return self.buffer[self._start_idx[index]: self._end_idx[index]].view(
+            self._tensors_sizes[index]
+        )
+
+    def __len__(self):
+        return self._tensors_len
+
+    def is_cuda(self):
+        return self.buffer.is_cuda
+
+    def nelement(self):
+        return self.buffer.nelement()
+
+    def unpack(self, tensors):
+        for tensor, entry in zip(tensors, self):
+            tensor.data[:] = entry
+
+
+def flatten(tensors, shapes=None, use_cuda=True):
```

Comment:
Can't this be replaced by `return torch.cat([p.data.view(-1) for p in tensors])`, which is same as https://github.com/pytorch/pytorch/blob/ce4f71c2b539cd76b094fa8fff75d670aec93df7/torch/distributed/algorithms/model_averaging/utils.py#L26?